src/nunit.analyzers/ValueSourceUsage/ValueSourceUsageAnalyzer.cs

```diff
@@ -0,0 +1,171 @@
+
+            if (syntaxNode == null || stringConstant == null)
+            {
+                return;
+            }
+
+            var symbol = SourceHelpers.GetMember(context, attributeInfo);
+            if (symbol is null)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(
+                    missingSourceDescriptor,
+                    syntaxNode.GetLocation(),
+                    stringConstant));
+            }
+            else
+            {
+                var sourceIsAccessible = context.SemanticModel.IsAccessible(
+                    syntaxNode.GetLocation().SourceSpan.Start,
+                    symbol);
+
+                if (attributeInfo.IsStringLiteral && sourceIsAccessible)
+                {
+                    var nameOfClassTarget = attributeInfo.SourceType.ToMinimalDisplayString(
+                        context.SemanticModel,
+                        syntaxNode.GetLocation().SourceSpan.Start);
+
+                    var nameOfTarget = attributeInfo.SourceType == context.ContainingSymbol.ContainingType
+                        ? stringConstant
+                        : $"{nameOfClassTarget}.{stringConstant}";
+
+                    var properties = new Dictionary<string, string>
+                    {
+                        { SourceCommonConstants.PropertyKeyNameOfTarget, nameOfTarget }
+                    };
+
+                    context.ReportDiagnostic(Diagnostic.Create(
+                        considerNameOfDescriptor,
+                        syntaxNode.GetLocation(),
+                        properties.ToImmutableDictionary(),
+                        nameOfTarget,
+                        stringConstant));
+                }
+
+                if (!symbol.IsStatic)
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(
+                        sourceNotStaticDescriptor,
+                        syntaxNode.GetLocation(),
+                        stringConstant));
+                }
+
+                switch (symbol)
+                {
+                    case IPropertySymbol property:
+                        ReportIfSymbolNotIEnumerable(context, syntaxNode, property.Type);
+                        break;
+                    case IFieldSymbol field:
+                        ReportIfSymbolNotIEnumerable(context, syntaxNode, field.Type);
+                        break;
+                    case IMethodSymbol method:
+                        ReportIfSymbolNotIEnumerable(context, syntaxNode, method.ReturnType);
+
+                        if (method.Parameters.Length != 0)
+                        {
+                            context.ReportDiagnostic(Diagnostic.Create(
+                                methodExpectParameters,
+                                syntaxNode.GetLocation(),
+                                method.Parameters.Length));
+                        }
+                        break;
+                }
```

Comment:
(feel free to ignore)

```suggestion                
                var memberType = symbol switch
                {
                    IPropertySymbol property => property.Type,
                    IFieldSymbol field => field.Type,
                    IMethodSymbol method => method.ReturnType,
                    _ => null
                };

                if (memberType != null)
                {
                    ReportIfSymbolNotIEnumerable(context, syntaxNode, memberType);
                }

                if(symbol is IMethodSymbol methodSymbol && methodSymbol.Parameters.Length != 0)
                {
                    context.ReportDiagnostic(Diagnostic.Create(
                        methodExpectParameters,
                        syntaxNode.GetLocation(),
                        methodSymbol.Parameters.Length));
                }
```

Reply:
Additional suggestion - to move the `memberType` switch to `ReportIfSymbolNotIEnumerable` method - to make `AnalyzeAttribute` a bit more concise.

Reply:
I actually tried to something similar, but it did not occur to me that I could use switch expressions (don't think that I've ever use them). I like the change 👍